### DISCUSSION TITLE
node 16-buster slim >> getting SSL error

### DISCUSSION THREAD

**Discussion by PKASHYAP-BEC** [iTwin/itwinjs-core#4508] [Q&A]
We are getting SSL error(download manifest failed (1) - SSL connect error) in BriefcaseManager.downloadBriefcase(args) method ,

do anyone have any solution to fix this issues , or any way to attach requested certs in all ITwin Calls.  
I have already tried below without any luck 
https://github.com/iTwin/itwinjs-core/discussions/2862

CODE Sample 
![image](https://user-images.githubusercontent.com/60220061/196698447-83ac93b6-f8ab-4d4e-a581-24d7baf74281.png)
Actual Error 
![image](https://user-images.githubusercontent.com/60220061/196698605-c32253d9-dcfb-4ada-996d-47a46b7fa9a1.png)

**Comment by khanaffan** [comment]
@PKASHYAP-BEC  I am looking at this issue.

**Comment by khanaffan** [comment]
@PKASHYAP-BEC So I try to replicate your issue and could not reproduce it. I have use latest `itwinjs-core@3.4.1` for testing as you have not specified exact version you are using. If provide the exact version I can test with that.

My docker file looks as follows

```docker
FROM docker.io/library/node:16-buster-slim
ENV NODE_ENV production

RUN apt-get update && apt-get install -y --no-install-recommends ca-certificates
RUN update-ca-certificates

COPY . /app
WORKDIR  /app
RUN npm i
RUN node /app/lib/cjs/Main.js

```

My download code straight up tries to download the v2 checkpoint which is what download briefcase in your case is calling
```ts
async function execute() {
    Logger.initializeToConsole();
    Logger.setLevelDefault(LogLevel.Trace);    
    const iModelHost: IModelHostOptions = {};
    const iModelClient = new IModelsClient({ api: { baseUrl: `https://api.bentley.com/imodels` } });
    iModelHost.hubAccess = new BackendIModelsAccess(iModelClient);

    await IModelHost.startup(iModelHost);

    const accessToken = "*******";
    const localFile = "test.bim";
    const iTwinId = "*********";
    const iModelId = "*********;
    const changesetId = "*********";
      
    await V2CheckpointManager.downloadCheckpoint({
        localFile,
        onProgress: (loaded: number, total: number) => {
            console.log(`${loaded}/${total}`)
            return ProgressStatus.Continue;
        },
        checkpoint: {
            expectV2: true,
            accessToken,
            iTwinId,
            iModelId,
            changeset: { id: changesetId },
        }
    });
}
```
Which seems to work fine. The console.log shows file download progress and it completes fine without an error.

```log
Trace | core-backend.IModelHost | IModelHost.startup
0/0
0/0
0/0
0/0
0/0
0/0
8388608/33554432
8388608/33554432
20971520/33554432
20971520/33554432
29360128/33554432
33554432/33554432
COMMIT
--> cddc9cfee4c
cddc9cfee4c575ef993135f92daee843c9efc6183548fe382bb38f320a8f409b
```

**Reply by PKASHYAP-BEC** [reply]
Hi , I am checking with 3.4.21 version and having our setup on AKS where we are getting SSL error only in download briefcase method. 

I believe, some native modules are being used which is not picking our default cert path.

Can you tell me what library is being used in C++ module to make any API call ?

**Reply by kabentley** [reply]
the native HTTPS is done with [curl](https://curl.se/)

**Reply by PKASHYAP-BEC** [reply]
Sir , that is strange, curl is working but we are still facing SSL error when downloading briefcase.
any suggestion for troubleshoot or any method by which we enforce certs to be included

**Reply by nick4598** [reply]
Could you clarify the version that you're using? What packages are you using at version 3.4.21?

**Reply by PKASHYAP-BEC** [reply]
we am using @itwin/core-backend@3.2.0 and @bentley\imodeljs-native@3.4.21 come with dependency tree. 
also we are using node:buster image which have debian 10/buster as OS.

**Reply by nick4598** [reply]
When I install @itwin/core-backend@3.2.0, I end up with a imodeljs-native version of 3.2.8. Maybe your package.json has "@itwin/core-backend": "^3.2.0", and you're installing a newer version of @itwin/core-backend? Could you check your package-lock.json for "@itwin/core-backend" and see what version is in there?

**Reply by PKASHYAP-BEC** [reply]
Sir, as per your suggestion,  I checked and found you are right .we are having  @itwin/core-backend@3.4.3 and @bentley\imodeljs-native@3.4.21. Is it wrong??

**Reply by nick4598** [reply]
No that should be fine, I wasn't able to reproduce with @bentley/imodeljs-native@3.4.21 with a dockerfile based off of node 16-buster-slim though. When you said curl is working, how do you know? Could I see your dockerfile? Feel free to remove any sensitive information if necessary.

**Reply by PKASHYAP-BEC** [reply]
hi , below is my docker file, to check if we are able to connect with api.bentley.com , we had curl and when i get shell from from running docker container, curl https://api.bentley.com is not throwing any SSL error but resource not found response.

```
FROM node:16-buster
RUN apt-get update && apt-get install -y --no-install-recommends ca-certificates=20200601~deb10u2 curl=7.64.0-4+deb10u3 wget=1.20.1-1.1 tcpdump=4.9.3-1~deb10u2 && apt-get clean && rm -rf /var/lib/apt/lists/* && update-ca-certificates 
RUN chown -R node:node /tmp
WORKDIR /
COPY build-output/dist/app/lets-encrypt-r3.pem ./etc/ssl/certs/lets-encrypt-r3.pem
COPY build-output/dist/app/ca-certificates.crt ./etc/ssl/certs/ca-certificates.crt
ENV NODE_EXTRA_CA_CERTS=/etc/ssl/certs/ca-certificates.crt SSL_CERT_DIR=/etc/ssl/certs/
WORKDIR /Application
COPY build-output/dist .
WORKDIR /Application/app/server
USER node
EXPOSE 34001
ENTRYPOINT ["node","--max-http-header-size=16384","--report-uncaught-exception","--use-openssl-ca","--report-on-signal","--report-on-fatalerror","--report-dir=/tmp" ,"index.js"]
```

**Reply by PKASHYAP-BEC** [reply]
I have found that there is a https://imodelhubprodchv2sa01.blob.core.windows.net blob storage urls are being used when downloading the briefcases ,  which is blocked in our firewall . 
Do we have any documentation available for setting up firewall rules for using itwin services? if yes, where can i get those. 
can i have list of all available blob storage urls so that i can get those whitelisted in the firewall.

**Reply by nick4598** [reply]
Unfortunately, there is currently no documentation available about this topic. At the moment though, whitelisting that blob storage url should be sufficient. Were you able to test that out and did you have any issues with it?
